File: java/src/main/java/com/android/widget/PhotoViewer/widget/PhotoView.java
package com.android.widget.PhotoViewer.widget;

import android.content.Context;
import android.graphics.Matrix;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.GestureDetector;

import androidx.appcompat.widget.AppCompatImageView;

import com.android.widget.PhotoViewer.listener.OnMatrixChangedListener;
import com.android.widget.PhotoViewer.listener.OnOutsidePhotoTapListener;
import com.android.widget.PhotoViewer.listener.OnPhotoTapListener;
import com.android.widget.PhotoViewer.listener.OnScaleChangedListener;
import com.android.widget.PhotoViewer.listener.OnSingleFlingListener;
import com.android.widget.PhotoViewer.listener.OnViewDragListener;
import com.android.widget.PhotoViewer.listener.OnViewTapListener;

/**
 * Created by xuzhb on 2021/1/11
 * Desc:一个可以缩放的ImageView，具体用法参考 {@link PhotoViewAttacher}
 */
public class PhotoView extends AppCompatImageView {

    private PhotoViewAttacher attacher;
    private ScaleType pendingScaleType;

    public PhotoView(Context context) {
        this(context, null);
    }

    public PhotoView(Context context, AttributeSet attr) {
        this(context, attr, 0);
    }

    public PhotoView(Context context, AttributeSet attr, int defStyle) {
        super(context, attr, defStyle);
        init();
    }

    private void init() {
        attacher = new PhotoViewAttacher(this);
        super.setScaleType(ScaleType.MATRIX);
        if (pendingScaleType != null) {
            setScaleType(pendingScaleType);
            pendingScaleType = null;
        }
    }

    public PhotoViewAttacher getAttacher() {
        return attacher;
    }

    public void setOnMatrixChangeListener(OnMatrixChangedListener listener) {
        attacher.setOnMatrixChangeListener(listener);
    }

    public void setOnOutsidePhotoTapListener(OnOutsidePhotoTapListener listener) {
        attacher.setOnOutsidePhotoTapListener(listener);
    }

    public void setOnPhotoTapListener(OnPhotoTapListener listener) {
        attacher.setOnPhotoTapListener(listener);
    }

    public void setOnScaleChangeListener(OnScaleChangedListener onScaleChangedListener) {
        attacher.setOnScaleChangeListener(onScaleChangedListener);
    }

    public void setOnSingleFlingListener(OnSingleFlingListener onSingleFlingListener) {
        attacher.setOnSingleFlingListener(onSingleFlingListener);
    }

    public void setOnViewDragListener(OnViewDragListener listener) {
        attacher.setOnViewDragListener(listener);
    }

    public void setOnViewTapListener(OnViewTapListener listener) {
        attacher.setOnViewTapListener(listener);
    }

    @Override
    public void setOnLongClickListener(OnLongClickListener l) {
        attacher.setOnLongClickListener(l);
    }

    @Override
    public void setOnClickListener(OnClickListener l) {
        attacher.setOnClickListener(l);
    }

    public void setOnDoubleTapListener(GestureDetector.OnDoubleTapListener onDoubleTapListener) {
        attacher.setOnDoubleTapListener(onDoubleTapListener);
    }

    @Override
    public ScaleType getScaleType() {
        return attacher.getScaleType();
    }

    @Override
    public void setScaleType(ScaleType scaleType) {
        if (attacher == null) {
            pendingScaleType = scaleType;
        } else {
            attacher.setScaleType(scaleType);
        }
    }

    public void setRotationTo(float rotationDegree) {
        attacher.setRotationTo(rotationDegree);
    }

    public void setRotationBy(float rotationDegree) {
        attacher.setRotationBy(rotationDegree);
    }

    public float getMinimumScale() {
        return attacher.getMinimumScale();
    }

    public void setMinimumScale(float minimumScale) {
        attacher.setMinimumScale(minimumScale);
    }

    public float getMediumScale() {
        return attacher.getMediumScale();
    }

    public void setMediumScale(float mediumScale) {
        attacher.setMediumScale(mediumScale);
    }

    public float getMaximumScale() {
        return attacher.getMaximumScale();
    }

    public void setMaximumScale(float maximumScale) {
        attacher.setMaximumScale(maximumScale);
    }

    public void setScaleLevels(float minimumScale, float mediumScale, float maximumScale) {
        attacher.setScaleLevels(minimumScale, mediumScale, maximumScale);
    }

    public float getScale() {
        return attacher.getScale();
    }

    public void setScale(float scale) {
        attacher.setScale(scale);
    }

    public void setScale(float scale, boolean animate) {
        attacher.setScale(scale, animate);
    }

    public void setScale(float scale, float focalX, float focalY, boolean animate) {
        attacher.setScale(scale, focalX, focalY, animate);
    }

    public boolean isZoomable() {
        return attacher.isZoomable();
    }

    public void setZoomable(boolean zoomable) {
        attacher.setZoomable(zoomable);
    }

    public void setZoomTransitionDuration(int milliseconds) {
        attacher.setZoomTransitionDuration(milliseconds);
    }

    public void setAllowParentInterceptOnEdge(boolean allow) {
        attacher.setAllowParentInterceptOnEdge(allow);
    }

    public RectF getDisplayRect() {
        return attacher.getDisplayRect();
    }

    @Override
    public Matrix getImageMatrix() {
        return attacher.getImageMatrix();
    }

    public boolean setDisplayMatrix(Matrix finalRectangle) {
        return attacher.setDisplayMatrix(finalRectangle);
    }

    public void setDrawMatrix(Matrix matrix) {
        attacher.setDrawMatrix(matrix);
    }

    public void setSuppMatrix(Matrix matrix) {
        attacher.setSuppMatrix(matrix);
    }

    @Override
    public void setImageDrawable(Drawable drawable) {
        super.setImageDrawable(drawable);
        if (attacher != null) {
            attacher.update();
        }
    }

    @Override
    public void setImageResource(int resId) {
        super.setImageResource(resId);
        if (attacher != null) {
            attacher.update();
        }
    }

    @Override
    public void setImageURI(Uri uri) {
        super.setImageURI(uri);
        if (attacher != null) {
            attacher.update();
        }
    }

    @Override
    protected boolean setFrame(int l, int t, int r, int b) {
        boolean changed = super.setFrame(l, t, r, b);
        if (changed) {
            attacher.update();
        }
        return changed;
    }

}

File: kotlin/src/main/java/com/android/widget/PhotoViewer/widget/PhotoView.kt
package com.android.widget.PhotoViewer.widget

import android.content.Context
import android.graphics.Matrix
import android.graphics.RectF
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.AttributeSet
import android.view.GestureDetector
import androidx.annotation.AttrRes
import androidx.appcompat.widget.AppCompatImageView
import com.android.widget.PhotoViewer.listener.*

/**
 * Created by xuzhb on 2021/3/30
 * Desc:一个可以缩放的ImageView，具体用法参考PhotoViewAttacher
 */
class PhotoView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, @AttrRes defStyleAttr: Int = 0
) : AppCompatImageView(context, attrs, defStyleAttr) {

    private var mAttacher: PhotoViewAttacher = PhotoViewAttacher(this)
    private var mPendingScaleType: ScaleType? = null

    init {
        super.setScaleType(ScaleType.MATRIX)
        if (mPendingScaleType != null) {
            scaleType = mPendingScaleType!!
            mPendingScaleType = null
        }
    }

    fun getAttacher(): PhotoViewAttacher = mAttacher

    fun setOnMatrixChangeListener(listener: OnMatrixChangedListener) {
        mAttacher.setOnMatrixChangeListener(listener)
    }

    fun setOnOutsidePhotoTapListener(listener: OnOutsidePhotoTapListener) {
        mAttacher.setOnOutsidePhotoTapListener(listener)
    }

    fun setOnPhotoTapListener(listener: OnPhotoTapListener) {
        mAttacher.setOnPhotoTapListener(listener)
    }

    fun setOnScaleChangeListener(listener: OnScaleChangedListener) {
        mAttacher.setOnScaleChangeListener(listener)
    }

    fun setOnSingleFlingListener(listener: OnSingleFlingListener) {
        mAttacher.setOnSingleFlingListener(listener)
    }

    fun setOnViewDragListener(listener: OnViewDragListener) {
        mAttacher.setOnViewDragListener(listener)
    }

    fun setOnViewTapListener(listener: OnViewTapListener) {
        mAttacher.setOnViewTapListener(listener)
    }

    override fun setOnLongClickListener(listener: OnLongClickListener?) {
        listener?.let { mAttacher.setOnLongClickListener(it) }
    }

    override fun setOnClickListener(listener: OnClickListener?) {
        listener?.let { mAttacher.setOnClickListener(it) }
    }

    fun setOnDoubleTapListener(listener: GestureDetector.OnDoubleTapListener) {
        mAttacher.setOnDoubleTapListener(listener)
    }

    override fun getScaleType(): ScaleType {
        return mAttacher.getScaleType()
    }

    override fun setScaleType(scaleType: ScaleType) {
        mAttacher.setScaleType(scaleType)
    }

    fun setRotationTo(rotationDegree: Float) {
        mAttacher.setRotationTo(rotationDegree)
    }

    fun setRotationBy(rotationDegree: Float) {
        mAttacher.setRotationBy(rotationDegree)
    }

    fun getMinimumScale(): Float = mAttacher.getMinimumScale()

    fun setMinimumScale(minimumScale: Float) {
        mAttacher.setMinimumScale(minimumScale)
    }

    fun getMediumScale(): Float = mAttacher.getMediumScale()

    fun setMediumScale(mediumScale: Float) {
        mAttacher.setMediumScale(mediumScale)
    }

    fun getMaximumScale(): Float = mAttacher.getMaximumScale()

    fun setMaximumScale(maximumScale: Float) {
        mAttacher.setMaximumScale(maximumScale)
    }

    fun setScaleLevels(minimumScale: Float, mediumScale: Float, maximumScale: Float) {
        mAttacher.setScaleLevels(minimumScale, mediumScale, maximumScale)
    }

    fun getScale(): Float = mAttacher.getScale()

    fun setScale(scale: Float, animate: Boolean = false) {
        mAttacher.setScale(scale, animate)
    }

    fun setScale(scale: Float, focalX: Float, focalY: Float, animate: Boolean) {
        mAttacher.setScale(scale, focalX, focalY, animate)
    }

    fun isZoomable(): Boolean = mAttacher.isZoomable()

    fun setZoomable(zoomable: Boolean) {
        mAttacher.setZoomable(zoomable)
    }

    fun setZoomTransitionDuration(milliseconds: Int) {
        mAttacher.setZoomTransitionDuration(milliseconds)
    }

    fun setAllowParentInterceptOnEdge(allow: Boolean) {
        mAttacher.setAllowParentInterceptOnEdge(allow)
    }

    fun getDisplayRect(): RectF? = mAttacher.getDisplayRect()

    override fun getImageMatrix(): Matrix = mAttacher.getImageMatrix()

    fun setDisplayMatrix(finalRectangle: Matrix) = mAttacher.setDisplayMatrix(finalRectangle)

    fun setDrawMatrix(matrix: Matrix) {
        mAttacher.setDrawMatrix(matrix)
    }

    fun setSuppMatrix(matrix: Matrix) {
        mAttacher.setSuppMatrix(matrix)
    }

    override fun setImageDrawable(drawable: Drawable?) {
        super.setImageDrawable(drawable)
        mAttacher.update()
    }

    override fun setImageResource(resId: Int) {
        super.setImageResource(resId)
        mAttacher.update()
    }

    override fun setImageURI(uri: Uri?) {
        super.setImageURI(uri)
        mAttacher.update()
    }

    override fun setFrame(l: Int, t: Int, r: Int, b: Int): Boolean {
        val change = super.setFrame(l, t, r, b)
        if (change) {
            mAttacher.update()
        }
        return change
    }

}